### PR TITLE
Improve Float ceil/round/floor error messages

### PR DESCRIPTION
**Pull request by tsubery**
In the case where precision is out of valid range, the default message is FunctionClause error.
With this commit an ArgumentError will be raised when precision is out of valid range.

This change aligns the behavior with Integer.parse, Float.parse that return ArgumentError on invalid input.

### CODE REVIEW

File: lib/elixir/lib/float.ex [hunk 1]
```diff
@@ -106,11 +108,15 @@ defmodule Float do
...
   """
-  @spec floor(float, 0..15) :: float
-  def floor(number, precision \\ 0) when is_float(number) and precision in 0..15 do
+  @spec floor(float, precision_valid_range) :: float
+  def floor(number, precision \\ 0) when is_float(number) and precision in @precision_valid_range do
     round(number, precision, :floor)
   end
 
+  def floor(number, precision) when is_float(number) do
+    raise ArgumentError, "precision #{inspect precision} is out of valid range of #{inspect @precision_valid_range}"
```
Comment (adkron) [inline lib/elixir/lib/float.ex]: Since the same string is used many times I think it would be much nicer to create an exception type for this.

ie.
`raise InvalidPrecision, precision: precision, range: @precision_valid_range`

or even a private function.

Comment (adkron) [inline lib/elixir/lib/float.ex]: Since the same string is used many times I think it would be much nicer to create an exception type for this.

ie.
`raise InvalidPrecision, precision: precision, range: @precision_valid_range`

or even a private function.

**Comment by adkron** [comment]
The tests are looking for a capital letter in the error sting. The code has a lowercase. That is why the tests are red.

**Comment by tsubery** [comment]
@adkron, I extracted a function to raise the error. A special Exception seems to me too much for such an edge case. It also breaks the convention of other Integer/Float functions raising ArgumentError on invalid input.

**Comment by adkron** [comment]
Good work. :+1:

File: lib/elixir/lib/float.ex [hunk 2]
```diff
@@ -106,11 +108,15 @@ defmodule Float do
...
   """
-  @spec floor(float, 0..15) :: float
-  def floor(number, precision \\ 0) when is_float(number) and precision in 0..15 do
+  @spec floor(float, precision_valid_range) :: float
+  def floor(number, precision \\ 0) when is_float(number) and precision in @precision_valid_range do
     round(number, precision, :floor)
   end
 
+  def floor(number, precision) when is_float(number) do
+    raise_invalid_precision(precision)
```
Comment (lexmag) [inline lib/elixir/lib/float.ex]: We should `raise` here instead of inside helper function.
Otherwise `floor/2` could be removed from stacktrace in case of tail call optimisation.

File: lib/elixir/lib/float.ex [hunk 3]
```diff
@@ -143,11 +149,15 @@ defmodule Float do
...
   """
-  @spec ceil(float, 0..15) :: float
-  def ceil(number, precision \\ 0) when is_float(number) and precision in 0..15 do
+  @spec ceil(float, precision_valid_range) :: float
+  def ceil(number, precision \\ 0) when is_float(number) and precision in @precision_valid_range do
     round(number, precision, :ceil)
   end
 
+  def ceil(number, precision) when is_float(number) do
+    raise_invalid_precision(precision)
```
Comment (lexmag) [inline lib/elixir/lib/float.ex]: Same as https://github.com/elixir-lang/elixir/pull/5820/files#r103353722.

File: lib/elixir/lib/float.ex [hunk 4]
```diff
@@ -188,15 +198,19 @@ defmodule Float do
...
   # This implementation is slow since it relies on big integers.
   # Faster implementations are available on more recent papers
   # and could be implemented in the future.
-  def round(float, precision \\ 0) when is_float(float) and precision in 0..15 do
+  def round(float, precision \\ 0) when is_float(float) and precision in @precision_valid_range do
     round(float, precision, :half_up)
   end
 
+  def round(number, precision) when is_float(number) do
+    raise_invalid_precision(precision)
```
Comment (lexmag) [inline lib/elixir/lib/float.ex]: Same as https://github.com/elixir-lang/elixir/pull/5820/files#r103353722.

File: lib/elixir/lib/float.ex [hunk 5]
```diff
@@ -8,6 +8,8 @@ defmodule Float do
   import Bitwise
 
   @power_of_2_to_52 4503599627370496
+  @precision_valid_range 0..15
```
Comment (lexmag) [inline lib/elixir/lib/float.ex]: `_valid_` part feels redundant, I believe there are many cases where we simply omit it.

File: lib/elixir/lib/float.ex [hunk 6]
```diff
@@ -106,11 +108,15 @@ defmodule Float do
...
   """
-  @spec floor(float, 0..15) :: float
-  def floor(number, precision \\ 0) when is_float(number) and precision in 0..15 do
+  @spec floor(float, precision_range) :: float
+  def floor(number, precision \\ 0) when is_float(number) and precision in @precision_range do
     round(number, precision, :floor)
   end
 
+  def floor(number, precision) when is_float(number) do
+    raise ArgumentError, "precision #{inspect precision} is out of valid range of #{inspect @precision_range}"
```
Comment (josevalim) [inline lib/elixir/lib/float.ex]: You can refactor the message to a private function, it is important only for the raise call to be on this function. :)

Comment (tsubery) [inline lib/elixir/lib/float.ex]: Makes sense

**Comment by josevalim** [comment]
:heart: :green_heart: :blue_heart: :yellow_heart: :purple_heart:

**Comment by plicjo** [comment]
Nice, Gal! @tsubery
